File: frontend/lib/api/cover-letters.ts
import { API_URL, getAuthToken, request } from "./_client";
import type { CoverLetterDocument, ChatMessage, ChatExchange } from "@/types/resume";

export const coverLettersApi = {
  getCoverLetters: () =>
    request<{ cover_letters: CoverLetterDocument[]; total: number }>("/api/v1/cover-letters"),

  getCoverLetter: (id: string) =>
    request<CoverLetterDocument>(`/api/v1/cover-letters/${id}`),

  updateCoverLetter: (
    id: string,
    data: Partial<{
      title: string;
      content_json: Record<string, unknown>;
      latex_source: string;
      hiring_manager_name: string;
      hiring_manager_email: string;
      street_address: string;
      city: string;
      state_province: string;
      postal_code: string;
      letter_date: string;
      additional_context: string;
    }>
  ) =>
    request<CoverLetterDocument>(`/api/v1/cover-letters/${id}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    }),

  createCoverLetter: (data: {
    title?: string;
    resume_id: string;
    hiring_manager_name?: string;
    hiring_manager_email?: string;
    street_address?: string;
    city?: string;
    state_province?: string;
    postal_code?: string;
    letter_date?: string;
    additional_context?: string;
  }) =>
    request<CoverLetterDocument>("/api/v1/cover-letters", { method: "POST", body: JSON.stringify(data) }),

  regenerateCoverLetter: (id: string) =>
    request<CoverLetterDocument>(`/api/v1/cover-letters/${id}/regenerate`, { method: "POST" }),

  getCoverLetterPreviewHtml: async (id: string): Promise<string> => {
    const token = getAuthToken();
    const res = await fetch(`${API_URL}/api/v1/cover-letters/${id}/preview`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });
    return res.text();
  },

  getCoverLetterMessages: (id: string) =>
    request<ChatMessage[]>(`/api/v1/cover-letters/${id}/messages`),

  sendCoverLetterChat: (id: string, message: string) =>
    request<ChatExchange>(`/api/v1/cover-letters/${id}/chat`, {
      method: "POST",
      body: JSON.stringify({ message }),
    }),

  handleCoverLetterChange: (id: string, change_id: string, action: "accept" | "reject") =>
    request<{ ok: boolean; content_json: Record<string, unknown> }>(
      `/api/v1/cover-letters/${id}/changes`,
      { method: "POST", body: JSON.stringify({ change_id, action }) }
    ),

  downloadCoverLetterPdf: async (id: string): Promise<Blob> => {
    const token = getAuthToken();
    const res = await fetch(`${API_URL}/api/v1/cover-letters/${id}/pdf`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });
    if (!res.ok) throw new Error("PDF export failed");
    return res.blob();
  },
};

File: backend/app/api/schemas/cover_letters.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class CoverLetterCreate(BaseModel):
    title: str | None = None
    resume_id: UUID | None = None
    hiring_manager_name: str | None = None
    hiring_manager_email: str | None = None
    street_address: str | None = None
    city: str | None = None
    state_province: str | None = None
    postal_code: str | None = None
    letter_date: str | None = None
    additional_context: str | None = None
    content_json: dict | None = None


class CoverLetterUpdate(BaseModel):
    title: str | None = None
    content_json: dict | None = None
    latex_source: str | None = None
    hiring_manager_name: str | None = None
    hiring_manager_email: str | None = None
    street_address: str | None = None
    city: str | None = None
    state_province: str | None = None
    postal_code: str | None = None
    letter_date: str | None = None
    additional_context: str | None = None


class CoverLetterResponse(BaseModel):
    id: UUID
    title: str
    status: str
    resume_id: UUID | None
    hiring_manager_name: str | None
    hiring_manager_email: str | None
    street_address: str | None
    city: str | None
    state_province: str | None
    postal_code: str | None
    letter_date: str | None
    additional_context: str | None
    content_json: dict
    latex_source: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CoverLetterListResponse(BaseModel):
    cover_letters: list[CoverLetterResponse]
    total: int
